[CI][Github] Install Clang in Windows container

cmtice opened the pull request:
Downloads clang-for-windows from the LLVM releases website, decompresses and untars the images, and leave them in C:\clang\clang-msvc\...  Temporarily downloads the 'xz' utility to decompress the downloaded clang tarball image.

cmtice committed:
[CI][Github] Install Clang in Windows container

Downloads clang-for-windows from the LLVM releases website,
decompresses and untars the images, and leave them in
C:\clang\clang-msvc\...  Temporarily downloads the 'xz' utility
to decompress the downloaded clang tarball image.

boomanaiden154 commented on .github/workflows/containers/github-action-ci-windows/Dockerfile:
```diff
@@ -98,3 +98,37 @@ RUN powershell -Command \
+    # --- 1. Download and extract xz --- \
+    Set-Location C:\temp-download ; \
+    Write-Host "Downloading xz-utils..."; \
+    Invoke-WebRequest -Uri "http://github.com/tukaani-project/xz/releases/download/v5.8.1/xz-5.8.1-windows.zip" -OutFile "xz.zip"; \
```
Is it possible to verify a checksum here? Similar to what we do for sccache in the linux container: https://github.com/llvm/llvm-project/blob/93dd17a0e0a437f9c60450db52b019c9d40e3d05/.github/workflows/containers/github-action-ci/Dockerfile#L77

boomanaiden154 commented on .github/workflows/containers/github-action-ci-windows/Dockerfile:
```diff
@@ -98,3 +98,37 @@ RUN powershell -Command \
+    [System.IO.Compression.ZipFile]::ExtractToDirectory('C:\temp-download\xz.zip', 'C:\xz-utils'); \ 
+    # --- 2. Download and decompress Clang --- \
+    Write-Host "Downloading Clang..."; \
+    Invoke-WebRequest -Uri "http://github.com/llvm/llvm-project/releases/download/llvmorg-21.1.2/clang+llvm-21.1.2-x86_64-pc-windows-msvc.tar.xz" -OutFile "clang+llvm-21.1.2-x86_64-pc-windows-msvc.tar.xz" ; \
```
Can we pull the version number here into an `ENV` line? https://github.com/llvm/llvm-project/blob/93dd17a0e0a437f9c60450db52b019c9d40e3d05/.github/workflows/containers/github-action-ci/Dockerfile#L5

boomanaiden154 commented on .github/workflows/containers/github-action-ci-windows/Dockerfile:
```diff
@@ -98,3 +98,37 @@ RUN powershell -Command \
+
+RUN powershell -Command \
+    Set-Location C:\clang ; \
+    Rename-Item -Path "C:\clang\clang+llvm-21.1.2-x86_64-pc-windows-msvc" -NewName "C:\clang\clang-msvc" ;
```
This is the binary? Why are we renaming it to `clang-msvc`?

cmtice replied:
No, that's not the binary. That's the directory into which tar untar'd everything. I'm shortening the path name.  With this renaming, the clang-cl binary is in C:\clang\clang-msvc\bin\clang-cl.exe  I thought the very long path name would be a pain to type and a likely source of typos.

boomanaiden154 reviewed (commented):
How does the size of this image compare to the existing one? `docker image ls` should show unpacked image sizes.

boomanaiden154 commented:
> 26.1 GB vs 17.7 GB, I think.

As in the new one is ~9GB smaller? That doesn't make a lot of sense.

cmtice commented:
> > > 26.1 GB vs 17.7 GB, I think.
> > 
> > 
> > As in the new one is ~9GB smaller? That doesn't make a lot of sense.
> 
> No, the new one is ~9GB bigger. Sorry that wasn't clear in my previous answer. New image: 26.1 GB; old image: 17.7 GB.

Note: The downloaded clang includes a LOT more than just the clang-cl.exe file:

C:\>ls clang\clang-msvc
bin  include  lib  libexec  share

C:\>ls clang\clang-msvc\bin
LLVM-C.dll                    llvm-cxxmap.exe
LTO.dll                       llvm-debuginfo-analyzer.exe
Remarks.dll                   llvm-debuginfod-find.exe
amdgpu-arch.exe               llvm-debuginfod.exe
analyze-build                 llvm-diff.exe
bugpoint.exe                  llvm-dis.exe
c-index-test.exe              llvm-dlltool.exe
clang++.exe                   llvm-dwarfdump.exe
clang-apply-replacements.exe  llvm-dwarfutil.exe
clang-change-namespace.exe    llvm-dwp.exe
clang-check.exe               llvm-exegesis.exe
clang-cl.exe                  llvm-extract.exe
clang-cpp.exe                 llvm-gsymutil.exe
clang-doc.exe                 llvm-ifs.exe
clang-extdef-mapping.exe      llvm-install-name-tool.exe
clang-format.exe              llvm-jitlink.exe
clang-include-cleaner.exe     llvm-lib.exe
clang-include-fixer.exe       llvm-libtool-darwin.exe
clang-installapi.exe          llvm-link.exe
clang-linker-wrapper.exe      llvm-lipo.exe
clang-move.exe                llvm-lto.exe
clang-nvlink-wrapper.exe      llvm-lto2.exe
clang-offload-bundler.exe     llvm-mc.exe
clang-offload-packager.exe    llvm-mca.exe
clang-query.exe               llvm-ml.exe
clang-refactor.exe            llvm-ml64.exe
clang-reorder-fields.exe      llvm-modextract.exe
clang-repl.exe                llvm-mt.exe
clang-scan-deps.exe           llvm-nm.exe
clang-sycl-linker.exe         llvm-objcopy.exe
clang-tblgen.exe              llvm-objdump.exe
clang-tidy.exe                llvm-opt-report.exe
clang.exe                     llvm-otool.exe
clangd.exe                    llvm-pdbutil.exe
diagtool.exe                  llvm-profdata.exe
dsymutil.exe                  llvm-profgen.exe
find-all-symbols.exe          llvm-ranlib.exe
git-clang-format              llvm-rc.exe
git-clang-format.bat          llvm-readelf.exe
hmaptool                      llvm-readobj.exe
intercept-build               llvm-readtapi.exe
ld.lld.exe                    llvm-reduce.exe
ld64.lld.exe                  llvm-remarkutil.exe
libclang.dll                  llvm-rtdyld.exe
libiomp5md.dll                llvm-sim.exe
liblldb.dll                   llvm-size.exe
libomp.dll                    llvm-split.exe
llc.exe                       llvm-stress.exe
lld-link.exe                  llvm-strings.exe
lld.exe                       llvm-strip.exe
lldb-argdumper.exe            llvm-symbolizer.exe
lldb-dap.exe                  llvm-tblgen.exe
lldb-instr.exe                llvm-tli-checker.exe
lldb-server.exe               llvm-undname.exe
lldb.exe                      llvm-windres.exe
lli.exe                       llvm-xray.exe
llvm-addr2line.exe            modularize.exe
llvm-ar.exe                   nvptx-arch.exe
llvm-as.exe                   offload-arch.exe
llvm-bcanalyzer.exe           opt.exe
llvm-bitcode-strip.exe        pp-trace.exe
llvm-c-test.exe               reduce-chunk-list.exe
llvm-cat.exe                  run-clang-tidy
llvm-cfi-verify.exe           sancov.exe
llvm-cgdata.exe               sanstats.exe
llvm-config.exe               scan-build
llvm-cov.exe                  scan-build-py
llvm-ctxprof-util.exe         scan-build.bat
llvm-cvtres.exe               scan-view
llvm-cxxdump.exe              verify-uselistorder.exe
llvm-cxxfilt.exe              wasm-ld.exe

cmtice committed:
Address some reviewer comments:
   - Remove "Write-Host" lines
   - Add LLVM_VERSION environment variable
   - Remove unnecessary files from downloaded clang

cmtice committed:
Put cleaning up clang download in same layer as download event.

cmtice committed:
Add md5 checksum checks on the downloaded files.

boomanaiden154 commented on .github/workflows/containers/github-action-ci-windows/Dockerfile:
```diff
@@ -98,3 +98,45 @@ RUN powershell -Command \
     rm actions-runner-win.zip
+
+# Set the LLVM_VERSION environment variable
+RUN powershell -Command [Environment]::SetEnvironmentVariable('LLVM_VERSION', '21.1.2', 'Machine')
```
The `ENV` or `ARG` Dockerfile keywords don't work for setting this?

cmtice committed:
Use Docker ENV keyword to define LLVM_VERSION.
